### PR TITLE
[hanime] Add new extractor

### PR DESCRIPTION
**Pull request by BrutuZ**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Extractor for [hanime.tv](https://hanime.tv) (NSFW) using as many fields as they provide

### CODE REVIEW

**Comment by dstftw** [comment]
Read coding conventions.

**Comment by BrutuZ** [comment]
Did that cover all convention recommendations or I still missed anything?

File: youtube_dl/extractor/hanime.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                'ext': 'mp4',
+                'url': item_url,
+            }, {
+                'width': int_or_none(f.get('width')),
+                'height': int_or_none(f.get('height')),
+                'protocol': 'https',
+                'format_id': 'm3u8-%sp' % f.get('height'),
+                'format_note': '~8-50.00Kib',
+                'ext': 'm3u8',
+                'url': item_url,
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Remove.

File: youtube_dl/extractor/hanime.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        title = page_json.get('name') or api_json.get[0].get('video_stream_group_id')
+        tags = [t.get('text') for t in page_json.get('hentai_tags')]
+
+        formats = []
+        for f in api_json:
+            item_url = url_or_none(f.get('url')) or url_or_none('https://hanime.tv/api/v1/m3u8s/%s.m3u8' % f.get('id'))
+            format = [{
+                'width': int_or_none(f.get('width')),
+                'height': int_or_none(f.get('height')),
+                'filesize_approx': parse_filesize('%sMb' % f.get('filesize_mbs')),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: `float_or_none`.

File: youtube_dl/extractor/hanime.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                'protocol': 'm3u8',
+                'format_id': 'mp4-%sp' % f.get('height'),
+                'tbr': float_or_none(float_or_none(f.get('filesize_mbs'), invscale=8388), int_or_none(f.get('duration_in_ms'), 1000)),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Remove.

File: youtube_dl/extractor/hanime.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            }]
+            for i in format:
+                formats.append(i)
+        formats.reverse()
+
+        return {
+            'id': str_or_none(api_json[0].get('id')),
+            'display_id': video_slug,
+            'title': title,
+            'description': clean_html(page_json.get('description')).strip(),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Breaks.

File: youtube_dl/extractor/hanime.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                'format_note': '~8-50.00Kib',
+                'ext': 'm3u8',
+                'url': item_url,
+            }]
+            for i in format:
+                formats.append(i)
+        formats.reverse()
+
+        return {
+            'id': str_or_none(api_json[0].get('id')),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Breaks.

File: youtube_dl/extractor/hanime.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            'display_id': video_slug,
+            'title': title,
+            'description': clean_html(page_json.get('description')).strip(),
+            'thumbnails': [
+                {'preference': 0, 'id': 'Poster', 'url': url_or_none(page_json.get('poster_url'))},
+                {'preference': 1, 'id': 'Cover', 'url': url_or_none(page_json.get('cover_url'))},
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: DRY. url is mandatory.

File: youtube_dl/extractor/hanime.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            ],
+            'release_date': unified_strdate(page_json.get('released_at')),
+            'upload_date': unified_strdate(page_json.get('created_at')),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: timestamp.

File: youtube_dl/extractor/hanime.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            'creator': str_or_none(page_json.get('brand')),
+            'view_count': int_or_none(page_json.get('views')),
+            'like_count': int_or_none(page_json.get('likes')),
+            'dislike_count': int_or_none(page_json.get('dislikes')),
+            'duration': parse_duration('%sms' % f.get('duration_in_ms')),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: `float_or_none`.

File: youtube_dl/extractor/hanime.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            'tags': tags,
+            'censored': str_or_none(page_json.get('is_censored')),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: No such meta field.

File: youtube_dl/utils.py [hunk 10]
```diff
@@ -3524,7 +3524,7 @@ def int_or_none(v, scale=1, default=None, get_attr=None, invscale=1):
         return default
     try:
         return int(v) * invscale // scale
-    except (ValueError, TypeError):
+    except (ValueError, TypeError, ZeroDivisionError):
```
Comment (dstftw) [inline youtube_dl/utils.py]: No.

File: youtube_dl/utils.py [hunk 11]
```diff
@@ -3546,7 +3546,7 @@ def float_or_none(v, scale=1, invscale=1, default=None):
         return default
     try:
         return float(v) * invscale / scale
-    except (ValueError, TypeError):
+    except (ValueError, TypeError, ZeroDivisionError):
```
Comment (dstftw) [inline youtube_dl/utils.py]: No.

File: youtube_dl/extractor/hanime.py (hunk 9 above)
Comment (BrutuZ) [reply youtube_dl/extractor/hanime.py]: Is there any way I can pass this information to use with the output layout?

Comment (dstftw) [reply youtube_dl/extractor/hanime.py]: Depends on what does it mean.

Comment (BrutuZ) [reply youtube_dl/extractor/hanime.py]: Never mind, it is also a tag, which can't be used there either so makes there's no reason this one should.

File: youtube_dl/extractor/hanime.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+    def _real_extract(self, url):
+        video_slug = self._match_id(url)
+        page_json = self._html_search_regex(r'<script>.+__NUXT__=(.+?);<\/script>', self._download_webpage(url, video_slug), 'Inline JSON')
+        page_json = try_get(self._parse_json(page_json, video_slug), lambda x: x['state']['data']['video']['hentai_video'], dict) or {}
+        api_json = try_get(self._download_json(
+            'https://members.hanime.tv/api/v3/videos_manifests/%s' % video_slug,
+            video_slug,
+            'API Call', headers={'X-Directive': 'api'}), lambda x: x['videos_manifest']['servers'], list) or []
+        title = page_json.get('name')
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Mandatory.

File: youtube_dl/extractor/hanime.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        duration = parse_duration('%sms' % page_json.get('duration_in_ms'))
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Again: float_or_none, not parse_duration.

File: youtube_dl/extractor/hanime.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        tags = []
+        for tag in page_json.get('hentai_tags'):
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Breaks.

File: youtube_dl/extractor/hanime.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+        },
+        'params': {
+            # m3u8 download
+            'skip_download': True,
+        },
+    }
+
+    def _real_extract(self, url):
+        video_slug = self._match_id(url)
+        page_json = self._html_search_regex(r'<script>.+__NUXT__=(.+?);<\/script>', self._download_webpage(url, video_slug), 'Inline JSON')
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Nothing changed.

File: youtube_dl/extractor/hanime.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        for tag in page_json.get('hentai_tags'):
+            if tag.get('text'):
+                tags.append(tag.get('text'))
+        formats = []
+        video_id = None
+        for server in api_json:
+            for stream in server['streams']:
+                if stream.get('compatibility') != 'all':
+                    continue
+                item_url = sanitize_url(stream.get('url')) or sanitize_url('https://hanime.tv/api/v1/m3u8s/%s.m3u8' % stream.get('id'))
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Nothing changed.

File: youtube_dl/extractor/hanime.py [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+                width = int_or_none(stream.get('width'))
+                height = int_or_none(stream.get('height'))
+                format = {
+                    'width': width,
+                    'height': height,
+                    'filesize_approx': float_or_none(parse_filesize('%sMb' % stream.get('filesize_mbs'))),
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: See above.

File: youtube_dl/extractor/hanime.py [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+        formats.reverse()
+
+        return {
+            'id': video_id,
+            'display_id': video_slug,
+            'title': title,
+            'description': clean_html(page_json.get('description')),
+            'thumbnails': [
+                {'preference': 0, 'id': 'Poster', 'url': page_json.get('poster_url')},
+                {'preference': 1, 'id': 'Cover', 'url': page_json.get('cover_url')},
```
Comment (dstftw) [inline youtube_dl/extractor/hanime.py]: Nothing changed.

**Comment by BrutuZ** [comment]
Since I'm not an actual programmer, could I kindly ask to get more than a couple words on the requested changes? Laconic answers can (and some have) become a time-consuming guessing game 😕

**Pull request closed without merging**